core/base/v7/inc/ROOT/TDirectory.h

```diff
@@ -148,35 +148,31 @@ class TDirectory {
   /// \throws TDirectoryTypeMismatch if the object stored under this name is of
   ///   a type that is not a derived type of `T`.
   template <class T>
-  std::shared_ptr<T> Get(const std::string& name) {
-    ContentMap_t::iterator idx = fContent.find(name);
+  decltype(auto) Get(const std::string& name) {
+    auto idx = fContent.find(name);
     if (idx != fContent.end()) {
-      // FIXME: implement upcast!
-      if (auto dep
-        = dynamic_cast<Internal::TDirectoryEntryPtr<T>*>(idx->second.get())) {
-        return dep->GetPointer();
-      }
+      if (idx->second.GetTypeInfo() == typeid(ToContentType_t<T>))
```

Comment:
> Not all archs make unique_ptr-s identical across shared libraries

I don't understand about which `unique_ptr` you are speaking. I don't put `unique_ptr` to `unordered_map` anymore. Or, if you'r speaking about storing `unique_ptr` in `TDirectoryEntry`, I don't understand what is special about `unique_ptr` in comparison to other types.

Nevertheless, I don't see why we should write own comparison for `std::type_info`. I've examined libstdc++ and libc++ realizations of `std::type_info::operator==()`. Both realizations end up using `strcmp`.